server/src/main/java/org/apache/iotdb/db/metadata/MTree.java

```diff
@@ -50,131 +56,111 @@
+        cur.addChild(nodeName, new InternalMNode(nodeName, cur));
       }
-      cur.setDataFileName(levelPath);
+      cur.setStorageGroupName(storageGroupName);
       cur = cur.getChild(nodeName);
-      if (levelPath == null) {
-        levelPath = cur.getDataFileName();
+      if (storageGroupName == null) {
+        storageGroupName = cur.getStorageGroupName();
       }
       i++;
     }
-    cur.setDataFileName(levelPath);
-    return cur;
+    cur.setStorageGroupName(storageGroupName);
+    MNode leaf = new LeafMNode(nodeNames[nodeNames.length - 1], cur, dataType, encoding,
+        compressor);
+    leaf.getSchema().setProps(props);
+    leaf.setStorageGroupName(cur.getStorageGroupName());
+    cur.addChild(nodeNames[nodeNames.length - 1], leaf);
   }
 
-
   /**
-   * function for checking whether the given path exists.
+   * Add path to MTree. This is available IF and ONLY IF creating schema automatically is enabled
    *
-   * @param path -seriesPath not necessarily the whole seriesPath (possibly a prefix of a sequence)
+   * @param path device id
    */
-  boolean isPathExist(String path) {
-    String[] nodeNames = nodeNames = MetaUtils.getNodeNames(path, PATH_SEPARATOR);
+  MNode addPath(String path) throws MetadataException {
+    String[] nodeNames = MetaUtils.getNodeNames(path);
+    if (nodeNames.length <= 1 || !nodeNames[0].equals(root.getName())) {
+      throw new IllegalPathException(path);
+    }
     MNode cur = root;
-    int i = 0;
-    while (i < nodeNames.length - 1) {
-      String nodeName = nodeNames[i];
-      if (cur.getName().equals(nodeName)) {
-        i++;
-        nodeName = nodeNames[i];
-        if (cur.hasChild(nodeName)) {
-          cur = cur.getChild(nodeName);
-        } else {
-          return false;
-        }
-      } else {
-        return false;
+    for (int i = 1; i < nodeNames.length; i++) {
+      if (!cur.hasChildWithKey(nodeNames[i])) {
+        cur.addChild(nodeNames[i], new InternalMNode(nodeNames[i], cur));
       }
+      cur = cur.getChild(nodeNames[i]);
     }
-    return cur.getName().equals(nodeNames[i]);
+    return cur;
+  }
+
+  /**
+   * Check whether the given path exists
+   *
+   * @param path not necessarily the whole seriesPath (possibly a prefix of a sequence)
+   */
+  boolean isPathExist(String path) {
+    return isPathExist(dummyNode, path);
   }
 
   /**
-   * function for checking whether the given path exists under the given mNode.
+   * Check whether the given path exists under the given MNode
    */
-  boolean isPathExist(MNode node, String path) {
-    String[] nodeNames = MetaUtils.getNodeNames(path, PATH_SEPARATOR);
-    if (nodeNames.length < 1) {
-      return true;
-    }
-    if (!node.hasChild(nodeNames[0])) {
-      return false;
+  private boolean isPathExist(MNode node, String path) {
+    String[] nodeNames = MetaUtils.getNodeNames(path);
+    MNode cur;
+    if (node.equals(dummyNode)) {
```

Comment:
I find that isPathExist(MNode node, String path) is only used in isPathExist(String path). So, the node must be a dummyNode. Then, we could init the cur = root without this check. And I really do not get why we need a dummyNode. It's better to remove it.

Reply:
I'm not sure whether this method will be used in the future, so I retained this method. It could be removed currently.